src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs

```diff
@@ -279,7 +330,48 @@ protected override void DisposeInternal()
                 }
             });
 
+            if (_batches != null)
+            {
+                _batches.DisposeStreams = true;
+                _batches.Dispose();
+            }
+            _tempFileCache.Dispose();
+
             base.DisposeInternal();
         }
+
+        public class ReplicationBatches : IDisposable
+        {
+            private readonly ShardedIncomingReplicationHandler _parent;
+            public ReplicationBatch[] Batches;
+            public Dictionary<Slice, Stream> Streams = new();
+            public bool DisposeStreams;
+
+            public ReplicationBatches(ShardedIncomingReplicationHandler parent)
+            {
+                _parent = parent;
+                DisposeStreams = true;
```

Comment:
what is that?

Reply:
We can dispose the streams as long as we don't get `MissingAttachmentException`. 
If we get `MissingAttachmentException`, for example, for _"foo.png"_ stream while we are sending `"foo2.png"` (as in all new tests), **AND** we dispose the stream we sent in the last batch (`"foo2.png"`), we get a loop... in that case, we want to receive `"foo.png"` from the source and send it along with `"foo2.png"` on the same batch.